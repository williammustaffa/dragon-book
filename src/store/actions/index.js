// Session
export { default as userLogin } from "./user/userLogin";
export { default as userLoginSuccess } from "./user/userLoginSuccess";
export { default as userLoginFailure } from "./user/userLoginFailure";
export { default as userLogout } from "./user/userLogout";
export { default as userCheckSession } from "./user/userCheckSession";
export { default as userCheckSessionSuccess } from "./user/userCheckSessionSuccess";
export { default as userCheckSessionFailure } from "./user/userCheckSessionFailure";

// Profile
export { default as createProfile } from "./profiles/createProfile";
export { default as createProfileFailure } from "./profiles/createProfileFailure";
