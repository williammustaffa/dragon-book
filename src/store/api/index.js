
import database from "store/api/database";
import * as R from "ramda";
import { v1 as uuidv1 } from 'uuid';
import request from "./request";

const AUTH_TOKEN_KEY = "auth.token";

class API {
  /**
   * Fake API error delay
   * @param {String} value 
   * @param {Int} time 
   */
  static retrieveDataDelayed(value, time = 500) {
    return new Promise(resolve => {
      setTimeout(() => resolve(value), time);
    });
  }

  /**
   * Fake API error delay
   * @param {String} value 
   * @param {Int} time 
   */
  static retrieveErrorDelayed(value, time = 500) {
    return new Promise((resolve, reject) => {
      setTimeout(() => reject({ message: value }), time);
    });
  }

  /**
   * Create token for user
   * @param {*} param
   */
  static tokenize({ id, email }) {
    return btoa(JSON.stringify({ id, email }));
  }

  /**
   * Get user info from token
   */
  static getTokenizedInfo() {
    const token = localStorage.getItem(AUTH_TOKEN_KEY);
  
    return token && JSON.parse(atob(token));
  }

  /**
   * Parse response object
   * @param {*} res 
   */
  static parseReponse(res) {
    return res.data;
  }

  /**
   * Create a user profile
   */
  createProfile = (payload) => {
    const errors = [];

    const existingUserEmail = database.queryAll("profiles", {
      query: { email: payload.email }
    });

    const existingUserUsername = database.queryAll("profiles", {
      query: { username: payload.username }
    });

    if (existingUserEmail.length) {
      errors.push("The email address provided is already in use.");
    }

    if (existingUserUsername.length) {
      errors.push("The username provided is already in use.");
    }

    if (errors.length) {
      return API.retrieveErrorDelayed(errors.join("\r\n"));
    }

    database.insert("profiles", {
      id: uuidv1(),
      email: payload.email,
      username: payload.username,
      firstName: payload.firstName,
      lastName: payload.lastName,
      description: payload.description,
      password: payload.password,
      registrationDate: Date.now(),
    });

    database.commit();

    // Create user and login
    return this.userLogin(payload);
  }

  /**
   * Check user session an then populate data if needed
   */
  userCheckSession = () => {
    const data = API.getTokenizedInfo();

    const results = database.queryAll("profiles", {
      query: { id: data && data.id },
    });

    if (!results.length) {
      return API.retrieveErrorDelayed("No session found.");
    }

    const user = R.clone(results[0]);

    // No need to delay on first load
    return user;
  }

  /**
   * Do user login by given crendentials
   * @param {object} credentials 
   */
  userLogin({ email, password }) {
    const results = database.queryAll("profiles", {
      query: { email, password },
    });

    if (!results.length) {
      return API.retrieveErrorDelayed("You have entered an invalid username or password");
    }

    const user = R.clone(results[0]);

    // Fake a token storage
    localStorage.setItem(AUTH_TOKEN_KEY, API.tokenize(user));

    return API.retrieveDataDelayed(user, 3000);
  }

  /**
   * Clear user session when he loogs out
   */
  userLogout = () => {
    // Clear token
    localStorage.removeItem(AUTH_TOKEN_KEY);

    return API.retrieveDataDelayed(null, 0);
  }

  /**
   * List dragons
   */
  listDragons = () => {
    return request
      .get("/dragon")
      .then(API.parseReponse);
  }

  /**
   * Get dragon details
   * @param {String} id
   */
  getDragon = (id) => {
    return request
      .get(`/dragon/${id}`)
      .then(API.parseReponse);
  }

  /**
   * Get dragon details
   * @param {Object} data
   */
  createDragon = (data) => {
    return request
      .post("/dragon", data)
      .then(API.parseReponse);
  }

  /**
   * Update dragon details
   * @param {Object} data
   */
  updateDragon = (data) => {
    return request
      .put(`/dragon/${data.id}`, data)
      .then(API.parseReponse);
  }

  /**
   * Update dragon details
   * @param {String} data
   */
  deleteDragon = (id) => {
    return request
      .put(`/dragon/${id}`)
      .then(API.parseReponse);
  }
}

// Create a cached connector
let connectorInstance = null;

function getAPIConnector() {
  if (!connectorInstance) {
    connectorInstance = new API();
  }

  return connectorInstance;
}


export { API, getAPIConnector };