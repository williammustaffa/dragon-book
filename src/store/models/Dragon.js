import moment from "moment";

class Dragon {
  constructor(payload = {}) {
    const histories = payload.histories || [];

    this.id = payload.id || "";
    this.createdAt = payload.createdAt;
    this.name = payload.name || "";
    this.type = payload.type || "common";
    this.history = histories.join("/n");
    this.imageUrl = payload.imageUrl;
  }

  /**
   * Returns the dragon detail url
   */
  get detailsUrl() {
    return `/dragon/${this.id}`;
  }

  /**
   * Returns the dragon update url
   */
  get updateUrl() {
    return `/dragon/update/${this.id}`;
  }

  /**
   * Returns a formatted creation date
   * @returns {String} creation date
   */
  get creationDate() {
    return (
      this.createdAt &&
      moment(new Date(this.createdAt)).format("MMM Do YY")
    );
  }

  /**
   * Returns a color based on the dragon type
   * @returns {String} color
   */
  get typeColor() {
    return this.typeSpec.color;
  }

  /**
   * Returns dragon current type
   */
  get typeSpec() {
    return Dragon.types[this.type] || Dragon.types.common;
  }

  /**
   * Returns a computed short desciption
   * @returns {String} description
   */
  get shortDescription() {
    return this.history.substring(0, 200);
  }

  static types = {
    legendary: { id: "legendary", name: "Legendary", color: "yellow", stars: 4 },
    epic: { id: "epic", name: "Epic", color: "purple", stars: 3 },
    rare: { id: "rare", name: "Rare", color: "blue", stars: 2 },
    common: { id: "common", name: "Common", color: "black", stars: 1 },
  }

  /**
   * Initial reducer state
   */
  static state = {
    isFetching: false,
    details: new Dragon(),
    errorMessage: ""
  }
}

export default Dragon;