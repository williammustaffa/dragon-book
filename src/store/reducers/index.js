import { combineReducers } from "redux";
import { connectRouter } from "connected-react-router";
import dragons from "./dragons";
import dragon from "./dragon";
import user from "./user";

const createRootReducer = (history) => combineReducers({
  router: connectRouter(history),
  dragons,
  dragon,
  user,
});

export default createRootReducer;