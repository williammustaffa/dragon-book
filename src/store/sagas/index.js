import { all } from "redux-saga/effects";

// User
import { watchUserLoginAsync } from "./user/userLogin";
import { watchUserLogoutAsync } from "./user/userLogout";
import { watchUserCheckSessionAsync } from "./user/userCheckSession";
import { watchCreateProfileAsync } from "./user/createProfile";

// Dragons
import { watchFetchDragons } from "./dragons/fetchDragons";
import { watchFetchDragon } from "./dragon/fetchDragon";

/**
 * Root saga provided to redux middleware
 */
function* root() {
  const watchers = [];

  // User watchers
  watchers.push(watchUserLoginAsync());
  watchers.push(watchUserLogoutAsync());
  watchers.push(watchUserCheckSessionAsync());
  watchers.push(watchCreateProfileAsync());

  // Dragons watchers
  watchers.push(watchFetchDragon());
  watchers.push(watchFetchDragons());

  yield all(watchers);
}

export default root;