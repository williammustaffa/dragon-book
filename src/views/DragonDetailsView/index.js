export { default as DragonDetailsView } from "./DragonDetailsView"