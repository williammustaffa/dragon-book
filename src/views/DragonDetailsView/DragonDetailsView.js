import React, { useEffect } from "react";
import { Grid, Image, Header, Divider, Button, Icon } from "semantic-ui-react";
import { useSelector, useDispatch } from "react-redux";
import { fetchDragon } from "store/actions";
import { Spinner } from "components/Spinner";
import { ErrorMessage } from "components/ErrorMessage";
import { push } from "connected-react-router";
import imageNotAvailable from "assets/images/imageNotAvailable.png";

function DragonDetailsView(props) {
  const { id } = props.match.params;
  const dispatch = useDispatch();
  const navigateTo = path => () => dispatch(push(path));

  useEffect(function () {
    dispatch(fetchDragon(id));
  }, [dispatch, id]);

  const { dragon, isFetching, errorMessage } = useSelector(function (state) {
    return {
      dragon: state.dragon.details,
      isFetching: state.dragon.isFetching,
      errorMessage: state.dragon.errorMessage 
    }
  });

  // Handle loading
  if (isFetching) {
    return <Spinner />;
  }

  // Handle error
  if (errorMessage) {
    return <ErrorMessage title="Error loading the page" content={errorMessage} />
  }

  return (
    <Grid>
      <Grid.Column mobile={16} tablet={8} computer={4}>
        <Image
          src={dragon.imageUrl || imageNotAvailable}
          label={{
            as: 'a',
            color: dragon.typeColor,
            content: dragon.type,
            attached: "bottom",
            className: "text-right text-uppercase"
          }}
        />
        <Button
          color="red"
          fluid
          onClick={navigateTo(dragon.updateUrl)}
          className="margin-top"
        >
          <Icon name="edit"/> Edit
        </Button>
      </Grid.Column>
      <Grid.Column mobile={16} tablet={8} computer={12}>
        <Header as="h2">{dragon.name}</Header>
        Created at {dragon.creationDate}
        <Divider />
        <p>{dragon.history}</p>
      </Grid.Column>
    </Grid>
  );
}

export default DragonDetailsView;