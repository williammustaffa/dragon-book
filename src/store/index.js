import { applyMiddleware, compose, createStore } from "redux";
import { routerMiddleware } from "connected-react-router";
import createRootReducer from "./reducers";
import sagas from "./sagas";
import history from "./history";
import createSagaMiddleware from "redux-saga"
import logger from "redux-logger";
import configureStore from "redux-mock-store";

/**
 * Generates a redux store properly
 * applying reducers, states and middlewares
 *
 * @param {*} preloadedState 
 */
function generateStore(preloadedState) {
  const sagaMiddleware = createSagaMiddleware();

  const store = createStore(
    createRootReducer(history),
    preloadedState,
    compose(
      applyMiddleware(
        routerMiddleware(history),
        sagaMiddleware,
        logger,
      ),
    ),
  );

  // Run redux saga
  sagaMiddleware.run(sagas);

  return store;
}

/**
 * Mock store and return a mocked state
 * @param {*} preloadedState 
 */
function mockStore(preloadedState) {
  const middlewares = [];
  const generateMockStore = configureStore(middlewares);

  return generateMockStore(preloadedState);
}

export {
  generateStore,
  mockStore,
  history,
};