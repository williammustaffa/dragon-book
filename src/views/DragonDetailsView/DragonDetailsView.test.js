import React from 'react';
import DragonDetailsView from './DragonDetailsView';
import { Provider } from 'react-redux';
import { mockStore } from "store";
import Dragon from 'store/models/Dragon';
import { render } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'

test('Test DragonDetailsView component', () => {
  // Create mock dragon object
  const dragon = Dragon.state;
  dragon.details = new Dragon({
    id: "1",
    name: "Saphira",
    type: "Epic",
    createdAt: "Sun May 10 2020 11:10:53 GMT-0300",
    histories: [],
  });

  // Mock store
  const store = mockStore({ dragon });

  const { asFragment } = render(
    <Provider store={store}>
      <DragonDetailsView match={{ params: { id: "1" } }} />
    </Provider>
  );

  // Run tests
  expect(asFragment()).toMatchSnapshot();
});