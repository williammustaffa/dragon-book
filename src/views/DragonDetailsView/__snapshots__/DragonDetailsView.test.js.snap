// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test DragonDetailsView component 1`] = `
<DocumentFragment>
  <div
    class="ui grid"
  >
    <div
      class="four wide computer sixteen wide mobile eight wide tablet column"
    >
      <div
        class="ui image"
      >
        <a
          class="ui black bottom attached label text-right text-uppercase"
        >
          Epic
        </a>
        <img
          src="imageNotAvailable.png"
        />
      </div>
      <button
        class="ui red fluid button margin-top"
      >
        <i
          aria-hidden="true"
          class="edit icon"
        />
         Edit
      </button>
    </div>
    <div
      class="twelve wide computer sixteen wide mobile eight wide tablet column"
    >
      <h2
        class="ui header"
      >
        Saphira
      </h2>
      Created at May 10th 20
      <div
        class="ui divider"
      />
      <p />
    </div>
  </div>
</DocumentFragment>
`;
